File: javascript/packages/ethereum/src/IsAHumanAttestationVerifier.js
import { EAS } from '@ethereum-attestation-service/eas-sdk';
import { ethers } from 'ethers';
import { createAttestationSuccess, createAttestationFailure } from '../../base/src/AttestationVerifier.js';
import { isRevoked, isExpired } from '../../base/src/RevocationExpirationHelper.js';
import { validateMerkleRootMatch } from './MerkleRootValidator.js';

const ZERO_REF_UID = '0x' + '0'.repeat(64);

function normalizeRefUid(refUid) {
  if (refUid == null) return ZERO_REF_UID;
  return typeof refUid === 'string'
    ? refUid.toLowerCase()
    : ethers.toBeHex(refUid, 32).toLowerCase();
}

function schemaUidToString(schemaUid) {
  if (schemaUid == null) return '';
  if (typeof schemaUid === 'string') return schemaUid.toLowerCase();
  try {
    return ethers.hexlify(schemaUid).toLowerCase();
  } catch {
    return String(schemaUid).toLowerCase();
  }
}

/**
 * IsAHumanAttestationVerifier verifies IsAHuman attestations (direct human identity).
 *
 * Supports two verification paths:
 * 1. Direct path: Attestation is the IsAHuman itself (refUID = 0)
 * 2. Indirect path: Attestation is IsAHuman with refUID pointing to a PrivateData or other subject
 *
 * When a valid attestation is found, sets:
 * - humanRootVerified: true
 * - humanVerification: { attester, rootSchemaUid }
 */
export class IsAHumanAttestationVerifier {
  /**
   * Create a new IsAHumanAttestationVerifier.
   *
   * @param {Map<string, Object>} networkConfig - Map of network ID to { rpcUrl, easContractAddress }
   * @param {Object} options - Optional configuration
   * @param {number} options.maxRefUIDDepth - Maximum depth for refUID following (default: 1)
   */
  constructor(networkConfig = new Map(), options = {}) {
    this.networkConfig = networkConfig;
    this.maxRefUIDDepth = options.maxRefUIDDepth ?? 1;
    this.serviceId = 'eas-is-a-human';
    this.easInstances = new Map();
  }

  /**
   * Legacy interface: verify without context.
   * Delegates to verifyWithContextAsync with empty context.
   *
   * @param {Object} attestation - Attestation from proof pack
   * @param {string} merkleRoot - Optional merkle root for validation
   * @returns {Promise<Object>} AttestationResult
   */
  async verifyAsync(attestation, merkleRoot) {
    return this.verifyWithContextAsync(attestation, { merkleRoot });
  }

  /**
   * Verify an IsAHuman attestation.
   *
   * @param {Object} attestation - Attestation from proof pack (has eas property)
   * @param {string} attestation.eas.network - Network name
   * @param {string} attestation.eas.attestationUid - Attestation UID
   * @param {Object} attestation.eas.schema - Schema info with schemaUid
   * @param {Object} context - Verification context with merkleRoot
   * @returns {Promise<Object>} AttestationResult with isValid, message, humanRootVerified, humanVerification
   */
  async verifyWithContextAsync(attestation, context) {
    const uidForResult = attestation?.eas?.attestationUid ?? 'unknown';

    if (!attestation?.eas) {
      const result = createAttestationFailure('Attestation or EAS data is null', 'MISSING_ATTESTATION', uidForResult);
      result.humanRootVerified = false;
      return result;
    }

    const { network, attestationUid, schema } = attestation.eas;

    if (!attestationUid) {
      const result = createAttestationFailure('Attestation UID is missing', 'INVALID_ATTESTATION_DATA', uidForResult);
      result.humanRootVerified = false;
      return result;
    }

    if (!network) {
      const result = createAttestationFailure('Network is missing', 'UNKNOWN_NETWORK', uidForResult);
      result.humanRootVerified = false;
      return result;
    }

    const eas = this.getEasInstance(network);
    if (!eas) {
      const result = createAttestationFailure(`Unknown network: ${network}`, 'UNKNOWN_NETWORK', attestationUid);
      result.humanRootVerified = false;
      return result;
    }

    try {
      const rootAttestation = await eas.getAttestation(attestationUid);
      if (!rootAttestation) {
        const result = createAttestationFailure('Root attestation not found on chain', 'ATTESTATION_DATA_NOT_FOUND', attestationUid);
        result.humanRootVerified = false;
        return result;
      }

      const locatorSchemaUid = schemaUidToString(schema?.schemaUid);
      const onChainSchemaUid = schemaUidToString(rootAttestation.schema);
      if (locatorSchemaUid && onChainSchemaUid && locatorSchemaUid !== onChainSchemaUid) {
        const result = createAttestationFailure(
          `On-chain schema ${onChainSchemaUid} does not match locator schema ${locatorSchemaUid}`,
          'SCHEMA_MISMATCH',
          attestationUid
        );
        result.humanRootVerified = false;
        return result;
      }

      if (isRevoked(rootAttestation)) {
        const result = createAttestationFailure('IsAHuman attestation is revoked', 'REVOKED', attestationUid);
        result.humanRootVerified = false;
        return result;
      }

      if (isExpired(rootAttestation)) {
        const result = createAttestationFailure('IsAHuman attestation is expired', 'EXPIRED', attestationUid);
        result.humanRootVerified = false;
        return result;
      }

      const rootSchemaUidString = onChainSchemaUid || locatorSchemaUid || '';

      const refUidNormalized = normalizeRefUid(rootAttestation.refUID);
      if (refUidNormalized !== ZERO_REF_UID) {
        const subjectResult = await this.verifySubjectAttestation(refUidNormalized, eas, context);

        if (!subjectResult.isValid) {
          const result = createAttestationFailure(
            subjectResult.message,
            subjectResult.reasonCode || 'VERIFICATION_ERROR',
            attestationUid
          );
          result.humanRootVerified = false;
          return result;
        }

        const success = createAttestationSuccess(
          'IsAHuman attestation verified (subject path)',
          attestationUid,
          'VALID',
          rootAttestation.attester ?? null
        );
        success.humanRootVerified = true;
        success.humanVerification = {
          attester: rootAttestation.attester ?? null,
          rootSchemaUid: rootSchemaUidString || null
        };
        return success;
      }

      const success = createAttestationSuccess(
        'IsAHuman attestation verified (direct path)',
        attestationUid,
        'VALID',
        rootAttestation.attester ?? null
      );
      success.humanRootVerified = true;
      success.humanVerification = {
        attester: rootAttestation.attester ?? null,
        rootSchemaUid: rootSchemaUidString || null
      };
      return success;
    } catch (error) {
      const result = createAttestationFailure(
        `Verification error: ${error.message}`,
        'VERIFICATION_ERROR',
        attestationUid
      );
      result.humanRootVerified = false;
      return result;
    }
  }

  /**
   * Verify a subject attestation (referenced by root's refUID).
   *
   * @private
   * @param {string} subjectUid - UID of the subject attestation
   * @param {EAS} eas - EAS instance
   * @param {Object} context - Verification context with merkleRoot
   * @returns {Promise<Object>} Result with isValid, message, reasonCode
   */
  async verifySubjectAttestation(subjectUid, eas, context) {
    try {
      const subjectAttestation = await eas.getAttestation(subjectUid);
      if (!subjectAttestation) {
        return {
          isValid: false,
          message: 'Subject attestation not found',
          reasonCode: 'ATTESTATION_DATA_NOT_FOUND'
        };
      }

      // Check if subject is revoked or expired using centralized helper
      if (isRevoked(subjectAttestation)) {
        return {
          isValid: false,
          message: 'Subject attestation is revoked',
          reasonCode: 'REVOKED'
        };
      }

      if (isExpired(subjectAttestation)) {
        return {
          isValid: false,
          message: 'Subject attestation is expired',
          reasonCode: 'EXPIRED'
        };
      }

      // If context has merkleRoot, verify it matches the subject data using centralized helper
      if (context?.merkleRoot && subjectAttestation.data) {
        const { isValid: merkleValid } = validateMerkleRootMatch(
          subjectAttestation.data,
          context.merkleRoot
        );

        if (!merkleValid) {
          return {
            isValid: false,
            message: 'Merkle root does not match subject attestation data',
            reasonCode: 'MERKLE_MISMATCH'
          };
        }
      }

      return { isValid: true };
    } catch (error) {
      return {
        isValid: false,
        message: `Subject verification error: ${error.message}`,
        reasonCode: 'VERIFICATION_ERROR'
      };
    }
  }

  /**
   * Get or create EAS instance for a network.
   *
   * @private
   * @param {string} networkId - Network ID
   * @returns {EAS|null} EAS instance or null if network not configured
   */
  getEasInstance(networkId) {
    if (this.easInstances.has(networkId)) {
      return this.easInstances.get(networkId);
    }

    const config = this.networkConfig.get(networkId);
    if (!config) {
      return null;
    }

    // For testing, allow injection of mock EAS
    if (config.easInstance) {
      this.easInstances.set(networkId, config.easInstance);
      return config.easInstance;
    }

    // Production path: create real EAS instance
    if (config.rpcUrl && config.easContractAddress) {
      try {
        const eas = new EAS(config.easContractAddress);
        eas.connect(config.rpcUrl);
        this.easInstances.set(networkId, eas);
        return eas;
      } catch {
        return null;
      }
    }

    return null;
  }
}

File: dotnet/src/Zipwire.ProofPack.Ethereum/ProofPack.Ethereum/IsAHumanAttestationVerifier.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Evoq.Blockchain;
using Evoq.Ethereum;
using Evoq.Ethereum.EAS;
using Evoq.Ethereum.JsonRPC;
using Microsoft.Extensions.Logging;
using Zipwire.ProofPack;

namespace Zipwire.ProofPack.Ethereum;

/// <summary>
/// Attestation verifier for human identity attestations (e.g., IsAHuman).
/// Validates that a human attestation is valid and optionally follows RefUID chains
/// to validate Merkle root binding through PrivateData or other schemas.
/// Implements both legacy IAttestationVerifier and context-aware IAttestationSpecialist.
/// </summary>
public class IsAHumanAttestationVerifier : IAttestationSpecialist
{
    private readonly Dictionary<string, EasNetworkConfiguration> _networkConfigurations;
    private readonly ILogger<IsAHumanAttestationVerifier>? _logger;
    private readonly Func<EasNetworkConfiguration, IGetAttestation> _getAttestationFactory;

    /// <summary>
    /// The service ID for this verifier.
    /// </summary>
    public string ServiceId => "eas-human";

    /// <summary>
    /// Initializes a new instance of IsAHumanAttestationVerifier.
    /// </summary>
    /// <param name="networkConfigurations">Network configurations for resolving EAS endpoints.</param>
    /// <param name="logger">Optional logger for debugging.</param>
    /// <param name="getAttestationFactory">Optional factory for creating IGetAttestation instances (for testing).</param>
    public IsAHumanAttestationVerifier(
        IEnumerable<EasNetworkConfiguration> networkConfigurations,
        ILogger<IsAHumanAttestationVerifier>? logger = null,
        Func<EasNetworkConfiguration, IGetAttestation>? getAttestationFactory = null)
    {
        this._networkConfigurations = networkConfigurations.ToDictionary(
            config => config.NetworkId,
            StringComparer.OrdinalIgnoreCase);
        this._logger = logger;
        this._getAttestationFactory = getAttestationFactory ?? CreateDefaultEasClient;
    }

    /// <inheritdoc />
    public async Task<AttestationResult> VerifyAsync(MerklePayloadAttestation attestation, Hex merkleRoot)
    {
        var context = new AttestationValidationContext(merkleRoot);
        return await VerifyAsyncWithContext(attestation, context);
    }

    /// <inheritdoc />
    public async Task<AttestationResult> VerifyAsyncWithContext(MerklePayloadAttestation attestation, AttestationValidationContext context)
    {
        // Validate input: null checks
        if (attestation?.Eas == null)
        {
            _logger?.LogWarning("Attestation or EAS data is null");
            return AttestationResult.Failure("Attestation or EAS data is null", AttestationReasonCodes.InvalidAttestationData, "unknown");
        }

        var easAttestation = attestation.Eas;

        // Resolve network config
        if (!_networkConfigurations.TryGetValue(easAttestation.Network, out var networkConfig))
        {
            _logger?.LogError("Unknown network: {Network}", easAttestation.Network);
            return AttestationResult.Failure(
                $"Unknown network: {easAttestation.Network}",
                AttestationReasonCodes.UnknownNetwork,
                easAttestation.AttestationUid);
        }

        try
        {
            _logger?.LogDebug("Verifying human attestation {AttestationUid} on network {Network}",
                easAttestation.AttestationUid, easAttestation.Network);

            var easClient = this._getAttestationFactory(networkConfig);
            var endpoint = networkConfig.CreateEndpoint();
            var interactionContext = new InteractionContext(endpoint, default);

            // Parse and validate attestation UID
            var attestationUid = easAttestation.AttestationUidHex;

            // Check if attestation exists and is valid
            var isValid = await easClient.IsAttestationValidAsync(interactionContext, attestationUid);
            if (!isValid)
            {
                _logger?.LogWarning("Attestation {AttestationUid} is not valid", easAttestation.AttestationUid);
                return AttestationResult.Failure(
                    $"Attestation {easAttestation.AttestationUid} is not valid",
                    AttestationReasonCodes.AttestationNotValid,
                    easAttestation.AttestationUid);
            }

            // Fetch full attestation data from EAS
            var fullAttestation = await easClient.GetAttestationAsync(interactionContext, attestationUid);
            if (fullAttestation == null)
            {
                _logger?.LogWarning("Could not fetch attestation {AttestationUid}", easAttestation.AttestationUid);
                return AttestationResult.Failure(
                    $"Could not fetch attestation {easAttestation.AttestationUid}",
                    AttestationReasonCodes.AttestationDataNotFound,
                    easAttestation.AttestationUid);
            }

            // Check revocation with detailed message
            var revocationCheck = RevocationExpirationHelper.CheckRevocation(fullAttestation);
            if (revocationCheck.IsRevoked)
            {
                _logger?.LogWarning("Attestation {AttestationUid} revocation check failed: {Message}", easAttestation.AttestationUid, revocationCheck.Message);
                return AttestationResult.Failure(
                    $"Attestation {easAttestation.AttestationUid} revocation check failed: {revocationCheck.Message}",
                    AttestationReasonCodes.Revoked,
                    easAttestation.AttestationUid);
            }

            // Check expiration
            var expirationCheck = RevocationExpirationHelper.CheckExpiration(fullAttestation);
            if (expirationCheck.IsRevoked)
            {
                _logger?.LogWarning("Attestation {AttestationUid} is expired: {Message}", easAttestation.AttestationUid, expirationCheck.Message);
                return AttestationResult.Failure(
                    $"Attestation {easAttestation.AttestationUid} is expired. {expirationCheck.Message}",
                    AttestationReasonCodes.Expired,
                    easAttestation.AttestationUid);
            }

            AttestationResult? innerRefResult = null;

            // Handle RefUID follow: use context.ValidateAsync when available so pipeline routes to the right verifier
            if (!fullAttestation.RefUID.IsZeroValue())
            {
                _logger?.LogDebug("Following RefUID {RefUID} from human attestation {AttestationUid}",
                    fullAttestation.RefUID, easAttestation.AttestationUid);

                var refUidHex = fullAttestation.RefUID;
                var referencedAttestation = await easClient.GetAttestationAsync(interactionContext, refUidHex);

                if (referencedAttestation == null)
                {
                    _logger?.LogWarning("Referenced attestation {RefUID} not found", fullAttestation.RefUID);
                    return AttestationResult.Failure(
                        $"Referenced attestation not found: {fullAttestation.RefUID}",
                        AttestationReasonCodes.MissingAttestation,
                        easAttestation.AttestationUid);
                }

                if (context.ValidateAsync != null)
                {
                    var refSchemaUid = referencedAttestation.Schema.ToString() ?? string.Empty;
                    if (string.IsNullOrEmpty(refSchemaUid))
                    {
                        _logger?.LogWarning("Referenced attestation {RefUID} has no schema", fullAttestation.RefUID);
                        return AttestationResult.Failure(
                            "Referenced attestation has no schema",
                            AttestationReasonCodes.UnknownSchema,
                            easAttestation.AttestationUid);
                    }

                    var refEas = new EasAttestation(
                        easAttestation.Network,
                        refUidHex.ToString(),
                        referencedAttestation.Attester.ToString(),
                        referencedAttestation.Recipient.ToString(),
                        new EasSchema(refSchemaUid, "Ref"));
                    var refAttestation = new MerklePayloadAttestation(refEas);
                    var refResult = await context.ValidateAsync(refAttestation);

                    if (!refResult.IsValid)
                    {
                        return AttestationResult.Failure(
                            $"Referenced attestation validation failed: {refResult.Message}",
                            refResult.ReasonCode ?? AttestationReasonCodes.VerificationError,
                            easAttestation.AttestationUid,
                            refResult);
                    }

                    innerRefResult = refResult;
                }
                else
                {
                    // Legacy path (no context.ValidateAsync): inline validation for ref
                    var refIsValid = await easClient.IsAttestationValidAsync(interactionContext, refUidHex);
                    if (!refIsValid)
                    {
                        _logger?.LogWarning("Referenced attestation {RefUID} is not valid", fullAttestation.RefUID);
                        return AttestationResult.Failure(
                            $"Referenced attestation is not valid: {fullAttestation.RefUID}",
                            AttestationReasonCodes.AttestationNotValid,
                            easAttestation.AttestationUid);
                    }

                    var refRevocationCheck = RevocationExpirationHelper.CheckRevocation(referencedAttestation);
                    if (refRevocationCheck.IsRevoked)
                    {
                        _logger?.LogWarning("Referenced attestation {RefUID} revocation check failed: {Message}", fullAttestation.RefUID, refRevocationCheck.Message);
                        return AttestationResult.Failure(
                            $"Referenced attestation {fullAttestation.RefUID} revocation check failed: {refRevocationCheck.Message}",
                            AttestationReasonCodes.Revoked,
                            easAttestation.AttestationUid);
                    }

                    var refExpirationCheck = RevocationExpirationHelper.CheckExpiration(referencedAttestation);
                    if (refExpirationCheck.IsRevoked)
                    {
                        _logger?.LogWarning("Referenced attestation {RefUID} is expired: {Message}", fullAttestation.RefUID, refExpirationCheck.Message);
                        return AttestationResult.Failure(
                            $"Referenced attestation is expired: {fullAttestation.RefUID}. {refExpirationCheck.Message}",
                            AttestationReasonCodes.Expired,
                            easAttestation.AttestationUid);
                    }

                    if (context.MerkleRoot.HasValue && referencedAttestation.Data != null && referencedAttestation.Data.Length > 0)
                    {
                        var merkleRootValue = context.MerkleRoot.Value;
                        var contextRootBytes = merkleRootValue.ToByteArray();
                        if (!referencedAttestation.Data.SequenceEqual(contextRootBytes))
                        {
                            _logger?.LogWarning("Merkle root mismatch in referenced attestation {RefUID}", fullAttestation.RefUID);
                            return AttestationResult.Failure(
                                "Merkle root does not match in referenced attestation",
                                AttestationReasonCodes.MerkleMismatch,
                                easAttestation.AttestationUid);
                        }
                    }
                }
            }

            // Success: return with HumanVerificationInfo and optional inner ref result
            var schemaUid = easAttestation.Schema?.SchemaUid;
            var humanVerification = new HumanVerificationInfo(
                verified: true,
                attester: fullAttestation.Attester.ToString(),
                rootSchemaUid: schemaUid);

            if (innerRefResult != null)
            {
                return new AttestationResult(
                    true,
                    $"Human attestation {easAttestation.AttestationUid} verified successfully",
                    "VALID",
                    easAttestation.AttestationUid,
                    fullAttestation.Attester.ToString(),
                    innerRefResult,
                    humanVerification);
            }

            return AttestationResult.Success(
                $"Human attestation {easAttestation.AttestationUid} verified successfully",
                fullAttestation.Attester.ToString(),
                easAttestation.AttestationUid,
                humanVerification);
        }
        catch (EasValidationException ex)
        {
            _logger?.LogError(ex, "EAS validation error while verifying attestation {AttestationUid}",
                easAttestation.AttestationUid);
            return AttestationResult.Failure(
                $"EAS validation error: {ex.Message}",
                AttestationReasonCodes.VerificationError,
                easAttestation.AttestationUid);
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Unexpected error while verifying human attestation {AttestationUid}",
                easAttestation.AttestationUid);
            return AttestationResult.Failure(
                $"Verification error: {ex.Message}\n{ex.StackTrace}",
                AttestationReasonCodes.VerificationException,
                easAttestation.AttestationUid);
        }
    }

    private static IGetAttestation CreateDefaultEasClient(EasNetworkConfiguration networkConfig)
    {
        var endpoint = networkConfig.CreateEndpoint();
        var address = networkConfig.EasContractAddress;
        return new ReadOnlyEasClient(address);
    }
}
